### PR TITLE
Enable PGO - alternative approach

### LINKED ISSUE #12636: dotnet fsi startup overhead surprisingly high
**Is your feature request related to a problem? Please describe.**

Startup time for `fsx` files is quite high.   It might not matter if you are planning to do a large computation, but for small scripts,  it seems to be routinely a few seconds, even for an empty fsx file with just `()` as code.   I ran some benchmarks against similar python programs for comparison on a few machines, [inspired by this comment]  (https://twitter.com/vilinski/status/1485188447639851012)
For small operating system tasks, running a tiny F# script would incur significant overhead, and it's a small barrier to starting up a quick REPL session/

Some tests I tried  

### Hello world, windows, dotnet 6.0.101, git bash

```
$ cat hw.fsx
printfn "hello world"

$ time dotnet fsi hw.fsx
hello world

real    0m2.711s
user    0m0.000s
sys     0m0.031s

daz@DarrenHP MINGW64 /c/src/tmp
$ dotnet --version
6.0.101
```

For comparison, here is hello world python

```
$ time python hw.py
hello world

real    0m0.386s
user    0m0.000s
sys     0m0.031s

$ cat hw.py
print("hello world")
$ python --version
Python 3.9.2

```

It's not even about really running the F# code itself,  similar tests get 1.8ish seconds for a completely empty .fsx file.  Interesting, adding in just the expression `()` adds about another half second.

```
$ cat null.fsx
()

$ time dotnet fsi null.fsx

real    0m2.397s
user    0m0.000s
sys     0m0.030s

$ cat empty.fsx

$ time dotnet fsi empty.fsx

real    0m1.778s
user    0m0.000s
sys     0m0.031s
```

Again python seems to shrug off having nothing to do very quickly

```
$ cat null.py

$ time python null.py

real    0m0.162s
user    0m0.000s
sys     0m0.031s

```

Curiously, and this is hard ti time but if I run fsi without a script associated, and just hit ctrl-D really really quickly,  I can be in and out of the reply in about a half-second (even with human reflexes), so there is something about an fsx file specifically which seems to slow things down. I guess that's good if you just want to play interactively

```
$ time dotnet fsi

Microsoft (R) F# Interactive version 12.0.0.0 for F# 6.0
Copyright (c) Microsoft Corporation. All Rights Reserved.

For help type #help;;

>
real    0m0.603s
user    0m0.000s
sys     0m0.031s
```

For comparison,  on a fairly high end linux AWS instance (m4.4xlarge, 16 vCPU, 64 GiB ram), with dotnet 6.0.101

```
time ~/.dotnet/dotnet fsi null.fsx

real    0m1.702s
user    0m1.588s
sys     0m0.131s
```

**Describe the solution you'd like**

It's totally a qualitative thing, but overhead for commencing execution of an fsx file should comparable to the startup file for the reply (let's say a half second but closer to the python times of 0.2 seconds would be ideal)

**Describe alternatives you've considered**

This doesn't matter if you are compiling code but for running scripts, especially lots of small scripts back to back, the two seconds would be problematic and making fewer smaller scripts or a different scripting system would be the solution.

**Additional context**

I did a little profiling with `strace` and `strace -c` on a linux machine.  I'm not sure this is super informative really - the total reported time in system calls is just 0.035 seconds.   I suspect the time is more in userland but I'd have to profile the whole startup to see that bit.  Trace output attached.  There does seem to be a large pause in the output from strace for a second or so in the middle just watching it interactively but I don't think it's necessarily making a system call at that time.

[strace-output.txt](https://github.com/dotnet/fsharp/files/7921301/strace-output.txt)
[strace-c-output.txt](https://github.com/dotnet/fsharp/files/7921302/strace-c-output.txt)

### LINKED ISSUE #13328: Add mibc/profile for R2R FSC/FSI for faster startup
We should consider adding both managed and native mibc/profiles for `fsc` and `fsi`:

- Come up with both FSC and FSI scenarios for profile generation, add them to `dotnet-optimization` (@EgorBo can assist with that):
	- Scenarios for collecting the profiling data (F# console, or web app, for example) which will be used to produce profiles.
	- Profiles then will be packaged and published as a NuGet package.
- When we building the compiler, download the NuGet package with profiles, and use them when building R2R compiler.

@EgorBo Have I missed anything?

### PR DESCRIPTION
**Pull request by psfinaki**
## Description

This should eventually make the FSC and FSI start about 30% faster. The implementation is inspired by the runtime targets.

Fixes #12636 
Fixes #13328

### CODE REVIEW

File: src/Directory.Build.props [hunk 1]
```diff
@@ -9,4 +9,64 @@
...
+    <TargetOS Condition="$([MSBuild]::IsOSPlatform('WINDOWS'))">windows</TargetOS>
+
+    <TargetArchitecture>$([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture.ToString().ToLowerInvariant)</TargetArchitecture>
+  </PropertyGroup>
+
+  <!-- MIBC data -->
+  <ItemGroup>
+
+    <!-- MIBC data to use when exact architecture match is available -->
+    <MIBCPackageDef
```
Comment (vzarytovskii) [inline src/Directory.Build.props:24]: Just fyi, these are coming from arcade already.

Comment (psfinaki) [reply src/Directory.Build.props:24]: So AFAIU no, they are not coming from arcade, these don't seem to be set currently. Also, I borrowed this from the runtime, where they'd probably rely on arcade if it was the case.

**Comment by psfinaki** [comment]
Hi @EgorBo, so this is an alternative approach to enable PGO, without doing profile generation ourselves.
However - this doesn't seem to work so far, as in, benchmarks don't show any improvements.

Any ideas what could be wrong?

**Comment by vzarytovskii** [comment]
@EgorBo any pointers on how to figure out why is profile we get from nuget is not applying, and if we generate it locally, it shows a difference?

**Comment by EgorBo** [comment]
@psfinaki @vzarytovskii I presume you test it on `fsi`? have you added `fsi` as a training scenario to dotnet-optimization repo? Otherwise it produces some MIBC trained to TechEmpower, console apps and for F# it's only "build Giraffe" scenario

**Comment by vzarytovskii** [comment]
> @psfinaki @vzarytovskii I presume you test it on `fsi`? have you added `fsi` as a training scenario to dotnet-optimization repo? Otherwise it produces some MIBC trained to TechEmpower, console apps and for F# it's only "build Giraffe" scenario

We have tested on `fsc`, since that's the only F# scenario in the optimization repo as of now. 

@psfinaki did you try to build Giraffe with and without profile applied in this PR? Does it change anything?

**Comment by psfinaki** [comment]
So far I have only tried running published `fsc` against an empty `test.fsx` script. That was enough to show the differece in that other PR with us creating the profile. 

Let me try Giraffe.

**Comment by psfinaki** [comment]
@vzarytovskii @EgorBo I'm afraid I don't see any difference in Giraffe compilation either :( it's the same execution time basically.

**Comment by EgorBo** [comment]
> @vzarytovskii @EgorBo I'm afraid I don't see any difference in Giraffe compilation either :( it's the same execution time basically.

Do you a MIBC that works locally? can you send it to me? I'll check weigths in it

**Comment by vzarytovskii** [comment]
@EgorBo did you, by any chance, looked at it? Curious what's the difference between those.

**Comment by EgorBo** [comment]
> @EgorBo did you, by any chance, looked at it? Curious what's the difference between those.

The only @psfinaki collected looks extremely weird to me. I propose we merge this PR as is since it generally looks corret to me, the problem must be inside the dotnet-optimization repo, once we figure it out and fix it should propagate a correct MIBC and optimize things out, hopefully 🙂 I'll take a look on Friday

**Review by EgorBo — APPROVED** [review]

File: src/fsc/fsc.targets [hunk 2]
```diff
@@ -54,4 +54,15 @@
     <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" />
   </ItemGroup>
 
+  <Target Name="Add PGO" DependsOnTargets="Get MIBC data" BeforeTargets="Publish" Condition="'$(Configuration)' == 'Release'">
```
Comment (KevinRansom) [inline src/fsc/fsc.targets:57]: We need to ship fsc and fsi to the dotnetsdk as il assemblies not native compiled assemblies.

The sdk native compiles F# in this project here: https://github.com/dotnet/sdk/blob/main/src/Layout/tool_fsharp/tool_fsc.csproj

Comment (KevinRansom) [reply src/fsc/fsc.targets:57]: So imagine, we will need to work with the sdk team to enable the pgo stuff ... at least if it relies on native compilation.

Comment (KevinRansom) [reply src/fsc/fsc.targets:57]: We called certainly add the mibc files to the package we insert into the sdk.

**Review by KevinRansom — CHANGES_REQUESTED** [review]
When I get less sick we should talk about this.

File: src/fsc/fsc.targets (hunk 2 above)
Comment (vzarytovskii) [reply src/fsc/fsc.targets:57]: I don't think sdk does anything, @baronfel doesn't it just repacks stuff from nugets?

Comment (psfinaki) [reply src/fsc/fsc.targets:57]: I was looking at the `tool_fsc` (and also [this](https://github.com/dotnet/sdk/blob/034362b0af690d938b7affe13cecc4eff29395b8/src/Layout/redist/targets/GenerateLayout.targets#L276-L284)) and that was my impression.

**Comment by psfinaki** [comment]
Will split this into 2 parts, closing this for now.

**Pull request closed without merging**